src/Control/Effect/Sum.hs

```diff
@@ -26,26 +26,48 @@ type Member sub sup = (Inject sub sup, Project sub sup)
 class Inject (sub :: (* -> *) -> (* -> *)) sup where
   inj :: sub m a -> sup m a
 
-instance Inject sub sub where
+instance Inject t t where
   inj = id
 
-instance {-# OVERLAPPABLE #-} Inject sub (sub :+: sup) where
-  inj = L . inj
-
-instance {-# OVERLAPPABLE #-} Inject sub sup => Inject sub (sub' :+: sup) where
+instance {-# OVERLAPPABLE #-}
+         Inject t (l1 :+: l2 :+: r)
+      => Inject t ((l1 :+: l2) :+: r) where
+  inj = reassoc . inj where
+    reassoc (L l)     = L (L l)
+    reassoc (R (L l)) = L (R l)
+    reassoc (R (R r)) = R r
```

Comment:
The reassociation here won’t be too bad a performance hit as long as we’re reasonable about using left-nested sums.

Reply:
Can you elaborate on what you mean by this?

Reply:
And can we document why we need to reassociate here (and maybe drop a link to DTALC for people who are wondering about why all these OVERLAPPABLE pragmas need to happen?

Reply:
> Can you elaborate on what you mean by this?
 
Basically, this assumes that trees will generally be right-leaning. So anything exclusively right-chained won’t ever see this instance, and if you have `(A :+: B) :+: C` that’s an extremely minor hit (as your benchmark demonstrates). But a long left-chain—which is _extremely_ unlikely to happen, and never by accident; you would have to work quite hard to design your effects thus—will take more of a hit because it has to do a bunch more reassociation.

Reply:
> And can we document why we need to reassociate here (and maybe drop a link to DTALC for people who are wondering about why all these OVERLAPPABLE pragmas need to happen?

I’m gonna do this in #223 to avoid conflicts.